pkg/apis/pipeline/v1alpha1/verificationpolicy_validation_test.go

```diff
@@ -95,7 +95,7 @@ func TestVerificationPolicy_Invalid(t *testing.T) {
 				},
 			},
 		},
-		want: apis.ErrMissingOneOf("key[0].data", "key[0].secretref"),
+		want: apis.ErrMissingOneOf("data", "kms", "secretref").ViaFieldIndex("key", 0),
```

Comment:
Can we make these test cases less fragile? If the given error changes in the implementation, this test will start failing. Maybe there's a way to untie them such that the test case is less fragile?
For example, what if someone has good reason to change the case from `kms` to `KMS` in the implementation -- as written, they now have to change the test code as well -- but this is not a functional change, it's just a change in the error message. Can we test for functional change instead of error message?

Reply:
This comment also applies to the ordering of the error message in index `0`. Can we be order-independent such that the test remains valid even if the order of validation changes in the implementation?

Reply:
This is how we return err for validation in current code base, take this for example: https://github.com/tektoncd/pipeline/blob/17b705c0efbc62741e41f4c99015db743552fb9b/pkg/apis/pipeline/v1beta1/pipelineref_validation.go#L37-L54
I think we should keep it consistent until we want to change them all

Reply:
I almost have the opposite take from Bendory-- I often give the feedback he is giving, but here I think it actually makes sense to include the text of the error message rather than `apis.ErrMultipleOneOf`. The reason is that sometimes these helper functions can lead to confusing error message strings, and you don't really realize what output the user is actually seeing until it's reflected in a test. This does have the downside of brittle tests, though.

Re: consistency, I think the section you're referring to is in non-test code, while the comment David is making is a separate point about brittleness of tests.

Reply:
> I almost have the opposite take from Bendory-- I often give the feedback he is giving, but here I think it actually makes sense to include the text of the error message rather than `apis.ErrMultipleOneOf`. The reason is that sometimes these helper functions can lead to confusing error message strings, and you don't really realize what output the user is actually seeing until it's reflected in a test. This does have the downside of brittle tests, though.
> 
> Re: consistency, I think the section you're referring to is in non-test code, while the comment David is making is a separate point about brittleness of tests.

Are you suggesting we should avoid using `apis.ErrMissingOneOf`?. It seems convenient and also used a lot in other validation functions. https://github.com/tektoncd/pipeline/search?q=apis.ErrMissingOneOf

Reply:
Only in tests. For example, in https://github.com/tektoncd/pipeline/issues/5673 I noted a confusing error message "expected exactly one, got both: spec.taskRef.name, spec.taskRef.params, spec.taskRef.resolver", which comes from `ErrMultipleOneOf`. These helpers are super useful in validation code but in tests IMO it's useful to be able to see the string error message the user encounters so you can check that it's not confusing.

The downside as David pointed out is that if you change "kms" in the code to "KMS", your tests will fail, but not in a way that gives you useful signal about the code. (The comment about indexing here I'm not sure I agree with, since the "0" refers to the index of the problematic field in the test data rather than an artifact of validation ordering.)

I may be going more in depth than is necessary; feel free to use your best judgment here although I'd just note our tests aren't actually that consistent ([example](https://github.com/tektoncd/pipeline/blob/b27ec733a7debf40a31f9628bc749603963b8fbb/pkg/apis/pipeline/v1beta1/task_validation_test.go#L789-L792))

Reply:
Ohh I see your point. Thanks!